File: examples/02_file/order.py
# Lean Code — BSD 3-Clause License — Vivian Voss, 2026
# Scope: Process an order into an invoice.

import sys


def read_order_from_file(path):
    lines = open(path).read().strip().split("\n")

    return {
        "id": int(lines[0]),
        "product": lines[1],
        "quantity": int(lines[2]),
    }


def read_stock_for_product(product):
    try:
        lines = open("stock.txt").read().strip().split("\n")
    except FileNotFoundError:
        return 0

    for line in lines:
        name, count = line.split(",", 1)
        if name == product:
            return int(count)

    return 0


def check_stock_availability(product, quantity):
    stock = read_stock_for_product(product)
    return stock >= quantity


def create_invoice_from_order(order, price_per_unit):
    return {
        "order_id": order["id"],
        "product": order["product"],
        "quantity": order["quantity"],
        "total": order["quantity"] * price_per_unit,
    }


def write_invoice_to_file(invoice, path):
    content = "\n".join([
        str(invoice["order_id"]),
        invoice["product"],
        str(invoice["quantity"]),
        f"{invoice['total']:.2f}",
    ])

    open(path, "w").write(content)


order = read_order_from_file("order.txt")

if not check_stock_availability(order["product"], order["quantity"]):
    sys.exit(f"out of stock: {order['product']}")

invoice = create_invoice_from_order(order, 29.99)
write_invoice_to_file(invoice, "invoice.txt")

File: examples/02_file/order.js
// Lean Code — BSD 3-Clause License — Vivian Voss, 2026
// Scope: Process an order into an invoice.

import { readFileSync, writeFileSync } from "node:fs";

function readOrderFromFile(path) {
  const lines = readFileSync(path, "utf-8").trim().split("\n");

  return {
    id: parseInt(lines[0], 10),
    product: lines[1],
    quantity: parseInt(lines[2], 10),
  };
}

function readStockForProduct(product) {
  const lines = readFileSync("stock.txt", "utf-8").trim().split("\n");

  for (const line of lines) {
    const [name, count] = line.split(",", 2);
    if (name === product) {
      return parseInt(count, 10);
    }
  }

  return 0;
}

function checkStockAvailability(product, quantity) {
  const stock = readStockForProduct(product);
  return stock >= quantity;
}

function createInvoiceFromOrder(order, pricePerUnit) {
  return {
    orderId: order.id,
    product: order.product,
    quantity: order.quantity,
    total: order.quantity * pricePerUnit,
  };
}

function writeInvoiceToFile(invoice, path) {
  const content = [
    invoice.orderId,
    invoice.product,
    invoice.quantity,
    invoice.total.toFixed(2),
  ].join("\n");

  writeFileSync(path, content);
}

const order = readOrderFromFile("order.txt");

if (!checkStockAvailability(order.product, order.quantity)) {
  process.stderr.write(`out of stock: ${order.product}\n`);
  process.exit(1);
}

const invoice = createInvoiceFromOrder(order, 29.99);
writeInvoiceToFile(invoice, "invoice.txt");
